---
title: "(m,n)-mer - A Simple and New Statistical Feature for viral Classification"
author:
- name: Amanda Araújo Serrão de Andrade
  affiliation: National Laboratory for Scientific Computing, Bioinformatics Laboratory (LABINFO), Petrópolis, Rio de Janeiro, Brazil
- name: Marco Grivet
  affiliation: Pontifícia Universidade Católica do Rio de Janeiro, 22451-900, Brazil
- name: Otávio Brustolini
  affiliation: National Laboratory for Scientific Computing, Bioinformatics Laboratory (LABINFO), Petrópolis, Rio de Janeiro, Brazil
- name: Ana Tereza Ribeiro de Vasconcelos
  affiliation: National Laboratory for Scientific Computing, Bioinformatics Laboratory (LABINFO), Petrópolis, Rio de Janeiro, Brazil
  email: atrv@lncc.br
package: mnmer
output:
  BiocStyle::html_document
abstract: |
  The (m,n)-mer is a new statistical feature based upon conditional frequencies (conditional probability density distributions). Here, we present the mnmer function and show a pratical example of classification using mnmer output.
vignette: |
  %\VignetteIndexEntry{(m,n)-mer - A Simple and New Statistical Feature for viral Classification}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[UTF-8]{inputenc}
---

\section{Introduction to the (m,n)-mer concept}

The (m,n)-mer R package was created to summarize biological data into numerical characteristics, as an alternative for k-mers. It reads a FASTA file and generates a table describing the conditional frequency distribution of the selected (m,n)-mer in the sequences. The feature matrix for classification is created by combining its output with class information.

Since letters are a bit awkward from a mathematical viewpoint, lets univocally associate the digits $0$, $1$, $2$ and $3$ to letters A, C, G and T. Any order will do. Hence, each k-mer can be described by a unique base-4 integer number in the range from $0$ to $4.k-1$. As an example, consider the 6-mer ACCTGA and the association described above. Then the 6-mer ACCTGA can be represented as the base-4 number $011320$, which corresponds to the number $376$ in the decimal notation. 

If we order all the k-mers according to these numbers, we say the k-mers are “lexicographically” ordered and we can use the notation $s_{k}^{i}$ to identify the k-mer associated to the decimal number $i$ ranging from $0$ to $4.k-1$. Following the above example, we can say that  $s^{376}_{6}=ACCTGA$.
Let’s now consider a particular genome extracted from a particular organism and assume that we count the occurrence of all distinct k-mers present in this genome. We will denote by $c_{k}^{i}$ the count corresponding to the i-th k-mer. Then we can create a real vector of size $4.k$ as illustrated in the Figure 1 below: 

\begin{figure*}
\includegraphics[keepaspectratio,width=12cm]{\jobname-Fig1.png}
\raggedright
\caption{\label{fig:wide}k-mer organization for k-mer and (m,n)-mer statistical distributions evaluation.}
\end{figure*}

Since this counting is generally made for several organisms with different sizes, for the sake of comparison it is more convenient express this count in relative terms. Similarly, we will denote by  as $f_{k}^{i}$ the relative frequency of k-mer $s_{k}^{i}$   which is computed as the division $f_{k}^{i}=\frac{c_{k}^{i}}{N_{k}}$   where ${N_{k}}$  is the total number of k-mers counted in this organism, that is, ${N_{k}}=\sum_{i=0}^{4^{k}-1}c_{k}^{i}$.

Let’s define a vector $\underline{f}_{k}=(f_{k}^{0},f_{k}^{1},...,f_{k}^{4^{k}-1})$   as the formal descriptor of the particular organism as far as k-mer is concerned. 

Please notice that each element of vector $\underline{f}_{k}$  is nonnegative and their sum is 1, which allow us to interpret this vector as a “probability density distribution” according to  statistical parlance.  

In the above example, consider the 6-mer ACCTGA and m and n respectively assuming the values 4 and 2. Hence we have $S_{6}=ACCTGA$ , $S_{4}^{-}=ACCT$ and $S_{2}^{+}=GA$ .
Although we have used the superscripts – and + to respectively indicate the left and right part of the k-mer  $S_{k}$, they are also an m-mer and an n-mer on its own.
We propose the replacement of the unconditional frequency $f_{k}^{i}$ by the conditional frequency  $f_{m,n}^{i}$ which represents the relative frequency of the n-mer $S_{n}^{+}$ conditioned to the fact that the set of m bases that precedes it is $S_{m}^{-}$. The vector  so defined will be conveniently renamed $\underline{f}_{m,n}$ in order to be more explicit. By defining  $u/v$ as the result of the integer division of $u$ by $v$ then, based upon the classical result of conditional probabilities, we have $\underline{f}_{m,n}=(f_{m,n}^{0},f_{m,n}^{1},...,f_{m,n}^{4^{k}-1})$  where $f_{m,n}^{i}=\frac{f_{k}^{i]}}{f_{m}^{i/4^n}}$.

Again in our example, $f_{6}^{376}$, the relative frequency of 6-mer  ACCTGA must be divided by $f_{4}^{23}$, which is the relative frequency of $S_{4}^{23}=ACCT$ .

It can be easily seen that the sum all elements of this new vector is no longer 1 but $4m$, because it is the concatenation of $4m$ conditional frequency distributions associated to each one of the possible m-mers. This fact has no impact whatsoever in our current discussion but, in order to keep it as a probability density distribution as well, we normalize it by dividing it by $4m$. Figure 2 shows an comparison of k-mers and (m,n)-mers obtained from the same nucleotide sequence. 

\begin{figure*}
\includegraphics[keepaspectratio,width=12cm]{\jobname-Fig2.png}
\raggedright
\caption{\label{fig:wide}Comparing k-mer to mn-mer relative frequency.}
\end{figure*}

According to Figure 3 below, the k-mers are represented as (0,k) and the (m,n)-mers as (m,n).

\begin{figure*}
\includegraphics[keepaspectratio,width=12cm]{\jobname-Fig3.png}
\raggedright
\caption{\label{fig:wide}Numeric representation.}
\end{figure*}

The output table (Figure 4) includes the fasta file accession numbers as an ID column, the relative frequency of mn-mers up to $4.k$ columns, and class information.

\begin{figure*}
\includegraphics[keepaspectratio,width=12cm]{\jobname-Fig4.png}
\raggedright
\caption{\label{fig:wide}Output example.}
\end{figure*}

For more details and performance comparison, please see Andrade et al., 2022 (in press).
\newpage

\section{The (m,n)-mer as an R package}
\begin{figure*}
\includegraphics[keepaspectratio,width=12cm]{\jobname-Fig5.png}
\raggedright
\caption{\label{fig:wide}Package logo}
\end{figure*}

The \Githubpkg{labinfo-lncc/mnmer} R package was created to summarize biological data into numerical characteristics. It accepts a FASTA file and generates a dataframe that describes the frequency of all (m,n)-mers identified in the input sequences. This output is coupled with class information to create a feature matrix. 

Limitations:

The software currently only handles simple FASTA files (no compression required);
In the (m,n)-mer generation, all bases other than A, C, T, and G are disregarded.
There is no parallelism or special C++ handling for large FASTA.
Prior to categorization, the user must add the data classes.

\subsection{Dependencies}

The package only needs \R{} 4.0.0 or later.

\subsection{Instalation}

The user should install the package from the GitHub repository. It can be done by using the \CRANpkg{devtools} package.

```{r, eval=FALSE}
library(devtools)
install_github("labinfo-lncc/mnmer", ref="main")
```

\subsection{The mnmer function}

The main function of this package is the `mnmer` function. It creates dataframes with the conditional probability.
By invoking the function  `cmmer` from the C++ script, this function can create both k-mers and (m,n)-mers.

The parameters receives:

`file` = FASTA file (it could be an multiFASTA)

`k` = Value of k for k-mer generation. Needs to be different from zero. 

`m` = Value of m for (m,n)-mer generation in the format of (m, k-m). In case of k-mer generation, m should be zero as (0,k).

\subsection{Pratical example}

The mnmer function generates a independent feature matrix that may be used to conduct clustering or classification. 

Assume we need to distinguish between viruses that only replicate in mosquito and viruses that only replicate in plants. The corresponding FASTA files can be found in the extdata folder.

After package installation, the user should run: 

```{r, eval=TRUE}
library("mnmer")
dir <-system.file("extdata", package="mnmer")
```

\subsubsection{Produce k-mer distributions}

For k-mer generation, the parameter k is set to choice, while the parameter m is set to zero. Given that the k-mers have been conditioned to zero bases.

```{r, eval=TRUE}
mosquito <- mnmer(file.path(dir, "mosquito_vir.fasta"),2,0)
plant <- mnmer(file.path(dir, "plant_vir.fasta"),2,0)
```

\subsubsection{Produce (m,n)-mer distributions}

The user specifies the k and m parameters for (mn)-mer generation.

For example, k = 2 and m = 1 produce the (1,1)-mer, in which one base is conditioned on the frequency of the base before it. Bases other than A, C, T, and G were disregarded.

```{r, eval=TRUE}
mosquito <- mnmer(file.path(dir, "mosquito_vir.fasta"),2,1)
plant <- mnmer(file.path(dir, "plant_vir.fasta"),2,1)
```

Here, we utilize the (1,1)-mer feature matrices generated by the mnmer to run an classification using Caret and MLeval.

Caret (https://topepo.github.io/caret/) is a library of functions for building predictive models from R systems. We utilized the createDataPartition, trainControl, and train functions in this example. The createDataPartition method splits the feature matrix and creates the train and test datasets. The trainControl function generates parameters that further regulate how the train function creates models.

MLeval (https://cran.r-project.org/web/packages/MLeval/index.html) is used to plot performance metrics. 

To execute the example, enter the following code:

```{r, eval=TRUE}
library(caret)
# Add class information 
classes <- replicate(nrow(mosquito), "mosquito.vir")
featureMatrix_mosquito <- cbind(mosquito,classes)
classes <- replicate(nrow(plant), "plant.vir")
featureMatrix_plant <- cbind(plant,classes)
featureMatrix <- rbind(featureMatrix_mosquito, featureMatrix_plant)
featureMatrix <- subset(featureMatrix, select = -c(seqid))

# Machine Learning
train_index <- caret::createDataPartition(featureMatrix$classes, p=0.8, list=FALSE)
train <- featureMatrix[train_index, ]
test <- featureMatrix[-train_index, ]
control <- caret::trainControl(method="cv", 
                        summaryFunction=twoClassSummary, 
                        classProbs=TRUE, 
                        savePredictions = TRUE)
roc <- caret::train(classes ~ ., 
            data=train, 
            method="rf", 
            preProc=c("center"), 
            trControl=control)
res <- MLeval::evalm(roc) # Make the ROC plot
```

\section{What to expect from newest versions}

Handle big multiFASTA files;

Paralelism;

Input compressed files; and

(m,n)-mer distribution for IUPACS and N bases; 

\section{SessionInfo}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
